dev/io.openliberty.jakartaee9.internal_fat/fat/src/io/openliberty/jakartaee9/internal/tests/EE9FeatureCompatibilityTest.java

```diff
@@ -0,0 +1,311 @@
+        for (String mpVersion : mpVersions) {
+            nonEE9MicroProfileFeatures.add("microProfile-" + mpVersion);
+        }
+
+        // MP standalone features
+        nonEE9MicroProfileFeatures.add("mpContextPropagation-1.0");
+        nonEE9MicroProfileFeatures.add("mpContextPropagation-1.2");
+        nonEE9MicroProfileFeatures.add("mpGraphQL-1.0");
+        nonEE9MicroProfileFeatures.add("mpLRA-1.0");
+        nonEE9MicroProfileFeatures.add("mpLRACoordinator-1.0");
+        nonEE9MicroProfileFeatures.add("mpReactiveMessaging-1.0");
+        nonEE9MicroProfileFeatures.add("mpReactiveStreams-1.0");
+
+        incompatibleValueAddFeatures.add("jwtSso-1.0"); // depends on mpJWT
+        incompatibleValueAddFeatures.add("openid-2.0"); // stabilized
+        incompatibleValueAddFeatures.add("openapi-3.1"); // depends on mpOpenAPI
+        incompatibleValueAddFeatures.add("opentracing-1.0"); // opentracing depends on mpConfig
+        incompatibleValueAddFeatures.add("opentracing-1.1");
+        incompatibleValueAddFeatures.add("opentracing-1.2");
+        incompatibleValueAddFeatures.add("opentracing-1.3");
+        incompatibleValueAddFeatures.add("opentracing-2.0");
+        incompatibleValueAddFeatures.add("sipServlet-1.1"); // purposely not supporting EE 9
+        incompatibleValueAddFeatures.add("springBoot-1.5"); // springBoot 3.0 will support EE 9
+        incompatibleValueAddFeatures.add("springBoot-2.0");
+
+        // The features set should contain all of the incompatible features.  If it doesn't
+        // something was removed or there is a typo.
+        for (String feature : nonEE9JavaEEFeatures) {
+            Assert.assertTrue(feature + " was not in the all features list", features.contains(feature));
+        }
+        for (String feature : nonEE9MicroProfileFeatures) {
+            Assert.assertTrue(feature + " was not in the all features list", features.contains(feature));
+        }
+        for (String feature : incompatibleValueAddFeatures) {
+            Assert.assertTrue(feature + " was not in the all features list", features.contains(feature));
+        }
+
+        File lib = new File(server.getInstallRoot(), "lib");
+
+        Utils.setInstallDir(lib.getParentFile());
+        KernelUtils.setBootStrapLibDir(lib);
+
+        BundleRepositoryRegistry.initializeDefaults(serverName, true);
+
+        repository = new FeatureRepository();
+        repository.init();
+    }
+
+    private static void parseShortName(File feature) throws IOException {
+        // Only scan *.mf files
+        if (feature.isDirectory() || !feature.getName().endsWith(".mf"))
+            return;
+
+        Scanner scanner = new Scanner(feature);
+        try {
+            String shortName = null;
+            while (scanner.hasNextLine()) {
+                String line = scanner.nextLine();
+                if (line.startsWith("IBM-ShortName:")) {
+                    shortName = line.substring("IBM-ShortName:".length()).trim();
```

Comment:
are we guaranteed to have IBM-ShortName header first?  Even for existing features that may not be using the `.feature` file?

Reply:
We are scanning the entire file.  Not sure why you are thinking this is looking for IBM-ShortName to be first.  We check every line  unless we get to a line that tells us this feature is one we do not want to include.